### PR TITLE
Implement LEAD, LAG, FIRST_VALUE, LAST_VALUE window functions

### LINKED ISSUE #2409: Window Offset Functions (LEAD/LAG/FIRST_VALUE/LAST_VALUE) Not Implemented
## Summary
Window offset functions `LEAD()`, `LAG()`, `FIRST_VALUE()`, and `LAST_VALUE()` are not implemented in AlaSQL. These are essential SQL:2003 standard window functions that allow accessing rows relative to the current row within a partition.

## Current Behavior

```javascript
// Attempting to use LEAD():
alasql('SELECT category, amount, LEAD(amount) OVER (PARTITION BY category ORDER BY amount) AS next_amt FROM data');

// Error: alasql.fn.LEAD is not a function
```

The same error occurs for `LAG()`, `FIRST_VALUE()`, and `LAST_VALUE()`.

### LEAD() - Access Next Row Value
```javascript
var data = [
  {category: 'A', amount: 10},
  {category: 'A', amount: 20},
  {category: 'A', amount: 30},
  {category: 'B', amount: 40}
];

alasql('SELECT category, amount, LEAD(amount) OVER (PARTITION BY category ORDER BY amount) AS next_amt FROM ?', [data]);

// Expected output:
[
  {category: 'A', amount: 10, next_amt: 20},   // Next in partition
  {category: 'A', amount: 20, next_amt: 30},   // Next in partition
  {category: 'A', amount: 30, next_amt: null}, // Last in partition
  {category: 'B', amount: 40, next_amt: null}  // Last in partition
]
```

### LAG() - Access Previous Row Value
```javascript
alasql('SELECT category, amount, LAG(amount) OVER (PARTITION BY category ORDER BY amount) AS prev_amt FROM ?', [data]);

// Expected output:
[
  {category: 'A', amount: 10, prev_amt: null}, // First in partition
  {category: 'A', amount: 20, prev_amt: 10},   // Previous in partition
  {category: 'A', amount: 30, prev_amt: 20},   // Previous in partition
  {category: 'B', amount: 40, prev_amt: null}  // First in partition
]
```

### FIRST_VALUE() - Access First Row in Window
```javascript
alasql('SELECT category, amount, FIRST_VALUE(amount) OVER (PARTITION BY category ORDER BY amount) AS first_amt FROM ?', [data]);

// Expected output:
[
  {category: 'A', amount: 10, first_amt: 10}, // First in partition
  {category: 'A', amount: 20, first_amt: 10}, // First in partition
  {category: 'A', amount: 30, first_amt: 10}, // First in partition
  {category: 'B', amount: 40, first_amt: 40}  // First in partition
]
```

### LAST_VALUE() - Access Last Row in Window
```javascript
alasql('SELECT category, amount, LAST_VALUE(amount) OVER (PARTITION BY category ORDER BY amount) AS last_amt FROM ?', [data]);

// Expected output (with proper frame specification):
[
  {category: 'A', amount: 10, last_amt: 30}, // Last in partition
  {category: 'A', amount: 20, last_amt: 30}, // Last in partition
  {category: 'A', amount: 30, last_amt: 30}, // Last in partition
  {category: 'B', amount: 40, last_amt: 40}  // Last in partition
]
```

### 1. Calculate Period-over-Period Change
```javascript
// Calculate month-over-month sales change
alasql(`
  SELECT 
    month, 
    sales,
    sales - LAG(sales) OVER (ORDER BY month) AS mom_change,
    (sales - LAG(sales) OVER (ORDER BY month)) / LAG(sales) OVER (ORDER BY month) * 100 AS mom_pct
  FROM monthly_sales
`);
```

### 2. Find Next Event Time
```javascript
// Find time until next customer visit
alasql(`
  SELECT 
    customer_id,
    visit_date,
    LEAD(visit_date) OVER (PARTITION BY customer_id ORDER BY visit_date) AS next_visit,
    DATEDIFF(day, visit_date, LEAD(visit_date) OVER (PARTITION BY customer_id ORDER BY visit_date)) AS days_until_next
  FROM customer_visits
`);
```

### 3. Compare with Baseline
```javascript
// Compare each employee salary to department min/max
alasql(`
  SELECT 
    dept,
    emp_name,
    salary,
    FIRST_VALUE(salary) OVER (PARTITION BY dept ORDER BY salary) AS dept_min,
    LAST_VALUE(salary) OVER (PARTITION BY dept ORDER BY salary ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS dept_max
  FROM employees
`);
```

### LEAD(expression [, offset [, default]]) OVER (...)
- `expression`: Column or expression to retrieve from the lead row
- `offset`: Number of rows forward (default: 1)
- `default`: Value to return when lead row doesn't exist (default: NULL)

### LAG(expression [, offset [, default]]) OVER (...)
- `expression`: Column or expression to retrieve from the lag row
- `offset`: Number of rows backward (default: 1)
- `default`: Value to return when lag row doesn't exist (default: NULL)

### FIRST_VALUE(expression) OVER (...)
- `expression`: Column or expression to retrieve from first row in window frame

### LAST_VALUE(expression) OVER (...)
- `expression`: Column or expression to retrieve from last row in window frame
- **Note**: Requires proper frame specification (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)

### 1. Parser Updates
Add function tokens to `src/alasqlparser.jison`:
```
LEAD | LAG | FIRST_VALUE | LAST_VALUE
```

### 2. Function Registration
Register functions in `src/55functions.js` or similar:
```javascript
stdlib.LEAD = function() { /* placeholder */ };
stdlib.LAG = function() { /* placeholder */ };
stdlib.FIRST_VALUE = function() { /* placeholder */ };
stdlib.LAST_VALUE = function() { /* placeholder */ };
```

### 3. Window Function Execution
Implement in `src/424select.js` or new `src/window-functions.js`:
- Detect offset functions during compilation
- After partitioning and ordering:
  - For LEAD: Access row at current_index + offset
  - For LAG: Access row at current_index - offset
  - For FIRST_VALUE: Access row at partition_start
  - For LAST_VALUE: Access row at partition_end (or frame_end)
- Handle default values when offset exceeds partition bounds

### 4. ORDER BY Requirement
These functions require ORDER BY in the OVER clause:
```javascript
// Should work:
LEAD(amount) OVER (PARTITION BY category ORDER BY date)

// Should error or warn:
LEAD(amount) OVER (PARTITION BY category)  // No ORDER BY
```

## Test Cases

```javascript
describe('Window Offset Functions', function() {
  var data = [
    {id: 1, category: 'A', value: 100, seq: 1},
    {id: 2, category: 'A', value: 200, seq: 2},
    {id: 3, category: 'A', value: 300, seq: 3},
    {id: 4, category: 'B', value: 150, seq: 1},
    {id: 5, category: 'B', value: 250, seq: 2}
  ];
  
  it('LEAD() with default offset of 1', function(done) {
    var res = alasql(
      'SELECT id, value, LEAD(value) OVER (PARTITION BY category ORDER BY seq) AS next_value FROM ?',
      [data]
    );
    assert.equal(res[0].next_value, 200);
    assert.equal(res[1].next_value, 300);
    assert.equal(res[2].next_value, null);
    done();
  });
  
  it('LAG() with custom offset', function(done) {
    var res = alasql(
      'SELECT id, value, LAG(value, 2) OVER (PARTITION BY category ORDER BY seq) AS prev_value_2 FROM ?',
      [data]
    );
    assert.equal(res[2].prev_value_2, 100); // Row 3 looks back 2 rows to row 1
    done();
  });
  
  it('LEAD() with default value', function(done) {
    var res = alasql(
      'SELECT id, value, LEAD(value, 1, 0) OVER (PARTITION BY category ORDER BY seq) AS next_value FROM ?',
      [data]
    );
    assert.equal(res[2].next_value, 0); // Default value instead of null
    done();
  });
  
  it('FIRST_VALUE() returns first in partition', function(done) {
    var res = alasql(
      'SELECT id, value, FIRST_VALUE(value) OVER (PARTITION BY category ORDER BY seq) AS first FROM ?',
      [data]
    );
    assert.equal(res[0].first, 100);
    assert.equal(res[1].first, 100);
    assert.equal(res[2].first, 100);
    done();
  });
  
  it('LAST_VALUE() with full frame', function(done) {
    var res = alasql(
      'SELECT id, value, LAST_VALUE(value) OVER (PARTITION BY category ORDER BY seq ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS last FROM ?',
      [data]
    );
    assert.equal(res[0].last, 300);
    assert.equal(res[1].last, 300);
    assert.equal(res[2].last, 300);
    done();
  });
});
```

## Implementation Priority
**Medium-High** - These are commonly used window functions in standard SQL and are expected by users familiar with PostgreSQL, SQL Server, and other modern databases.

## References
- SQL:2003 Standard - Window Functions
- PostgreSQL Window Functions: https://www.postgresql.org/docs/current/functions-window.html
- Existing implementation: `ROW_NUMBER()` can serve as a template for window function mechanics
- Parser already supports OVER clause structure (`src/47over.js`)

## Related Issues
- Issue #[original] - Request for window functions including LEAD/LAG
- Window aggregate functions issue (COUNT/MAX/MIN OVER)

### PR DESCRIPTION
**Pull request by apoorvdarshan**
## Summary
- Implements SQL:2003 positional window functions: `LEAD()`, `LAG()`, `FIRST_VALUE()`, `LAST_VALUE()` (resolves #2409)
- Supports `PARTITION BY` and `ORDER BY` in OVER clauses, with configurable offset and default values for LEAD/LAG
- Adds 16 test cases covering basic usage, partitioning, edge cases (nulls, offset > partition size, DESC ordering, multiple window fns)

## Test plan
- [x] All 16 new tests in `test/test2409.js` pass
- [x] Full regression suite passes (2457 passing, 0 failing)
- [x] Verify LEAD/LAG with explicit offset and custom default values (tests 2, 3, 6)
- [x] Verify FIRST_VALUE/LAST_VALUE with and without PARTITION BY (tests 8, 9, 11, 12)
- [x] Verify edge cases: nulls, single-row partitions, offset exceeding partition size (tests 13, 14)

### CODE REVIEW

**Comment by mathiasrw** [comment]
Are you intending to complete the test plan, or should I review now?

**Comment by apoorvdarshan** [comment]
Test plan complete — all items verified, ready for your review. Re-ran the full suite locally: 2457 passing, 0 failing, plus all 16 new tests in `test/test2409.js` pass. The three remaining manual checks are covered by the test cases (annotated in the updated description).

File: src/40select.js
Comment (mathiasrw) [inline src/40select.js:1]: Nice way to structure this functionality

File: src/424select.js [hunk 1]
```diff
@@ -570,6 +570,46 @@ yy.Select.prototype.compileSelectGroup0 = function (query) {
 			if (col.funcid && col.funcid.toUpperCase() === 'GROUP_ROW_NUMBER') {
 				query.grouprownums.push({as: col.as, columnIndex: 0}); // Track which column to use for grouping
 			}
+
+			// Detect positional window functions: LEAD, LAG, FIRST_VALUE, LAST_VALUE
```
Comment (mathiasrw) [inline src/424select.js]: Lets move the detection into the parser so we dont pollute the areas of concern. 

Please consider how we can add lead, lag, first value, last value to the grammar in [alasqlparser.jison](https://github.com/AlaSQL/alasql/blob/develop/src/alasqlparser.jison) (use `yarn jison` to build after changing the jison)

File: src/55functions.js [hunk 2]
```diff
@@ -250,6 +250,18 @@ stdlib.ROW_NUMBER = function () {
...
+};
+stdlib.LAG = function () {
+	return 'undefined';
+};
+stdlib.FIRST_VALUE = function () {
+	return 'undefined';
+};
+stdlib.LAST_VALUE = function () {
+	return 'undefined';
+};
```
Comment (mathiasrw) [inline src/55functions.js]: Its not great to have functions intended to not do anything other than being a flag for code to operate on. Lets move the functionality to the jison parser.

File: test/test2409.js [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+describe('Test 2409 - LEAD/LAG/FIRST_VALUE/LAST_VALUE Window Functions', function () {
+	var data = [
+		{dept: 'Sales', emp: 'Alice', salary: 1000},
+		{dept: 'Sales', emp: 'Bob', salary: 1200},
+		{dept: 'Sales', emp: 'Carol', salary: 1500},
+		{dept: 'IT', emp: 'Dave', salary: 2000},
+		{dept: 'IT', emp: 'Eve', salary: 2500},
+	];
+
+	// --- LEAD tests ---
```
Comment (mathiasrw) [inline test/test2409.js]: Please make sure the tests makes sense without comments

File: test/test2409.js [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+	it('1. LEAD basic - next row value', function () {
+		var res = alasql(
+			'SELECT emp, salary, LEAD(salary) OVER (ORDER BY salary) AS next_salary FROM ? ORDER BY salary',
+			[data]
+		);
+		assert.strictEqual(res[0].next_salary, 1200);
+		assert.strictEqual(res[1].next_salary, 1500);
+		assert.strictEqual(res[4].next_salary, null);
```
Comment (mathiasrw) [inline test/test2409.js]: Please seek to compare the output from alasql with the complete output json object so we provide the total output expected as one object.

File: test/test2409.js (hunk 3 above)
Comment (mathiasrw) [reply test/test2409.js]: I see now its because you are bundling across lead lag first last and so on. 

I propose making a level of `describe` for each of these so its clear in the code and ind the output. 

So instead of the comment on this line its something like 


```
describe('Test 2409 - LEAD/LAG/FIRST_VALUE/LAST_VALUE Window Functions', function () {
	describe('Test 2409 - LEAD Window Functions', function () { 
   		test(...)
	}

	describe('Test 2409 - LAG Window Functions', function () { 
   		...
	}
```

**Review by mathiasrw — CHANGES_REQUESTED** [review]
Nice work with making the solution work. But we should be focusing on using the jison grammar so we can use caching and pre compile instead of doing function detection at runtime.

**Comment by apoorvdarshan** [comment]
Addressed in d39981d. Summary of changes:

**Detection moved into the parser** (`src/alasqlparser.jison`)
- The `FuncValue` action now produces a typed `yy.PositionalWindowFunc` AST node when it sees `LEAD`/`LAG`/`FIRST_VALUE`/`LAST_VALUE`, so detection happens at parse time and the typed node is what gets cached/precompiled.
- `yy.PositionalWindowFunc` (defined in `src/47over.js`) carries the structured `args`, `over`, partition, and order info, and has its own `findAggregator` that registers the window-function config on the query — no more string-matching in `compileSelectGroup0`.
- Ran `yarn jison` to regenerate `src/alasqlparser.js`.

**Runtime detection block removed** (`src/424select.js`)
- The `if (col.funcid)` block that did string comparison and config-building at runtime is gone.

**No-op stdlib stubs removed** (`src/55functions.js`)
- The flag-only `stdlib.LEAD/LAG/FIRST_VALUE/LAST_VALUE` entries are gone; `yy.PositionalWindowFunc.toJS()` now returns `'undefined'` directly so the post-processor can fill the slot.

**Tests restructured** (`test/test2409.js`)
- Replaced the comment dividers with nested `describe` blocks (one per function).
- Replaced per-field `assert.strictEqual` calls with single `assert.deepStrictEqual` checks against the full expected result object.
- Removed the numeric prefixes/comments in test names.

Full suite: 2457 passing, 0 failing.

**Comment by apoorvdarshan** [comment]
Rebased onto current `develop` (new head d5a01d61). The requested parser-based detection, typed `PositionalWindowFunc` AST node, generated parser, per-function `describe` blocks, and whole-object assertions remain intact. Rebuilt successfully and reran `test/test2409.js`: 16 passing. Ready for re-review.

**Comment by apoorvdarshan** [comment]
@mathiasrw Friendly ping for re-review when you have a moment.

Since your April notes, LEAD/LAG/FIRST_VALUE/LAST_VALUE are dedicated jison grammar productions (typed AST, no runtime function-flag detection), tests use per-function `describe` blocks with full-object asserts, and the branch is rebased onto current `develop` (mergeable). Happy to adjust anything else you spot.
